akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala

```diff
@@ -1104,3 +1104,53 @@ private[stream] final class Reduce[T](f: (T, T) ⇒ T) extends SimpleLinearGraph
   }
   override def toString = "Reduce"
 }
+
+/**
+ * INTERNAL API
+ */
+private[stream] final class RecoverWith[T, M](pf: PartialFunction[Throwable, Graph[SourceShape[T], M]]) extends SimpleLinearGraphStage[T] {
+  override def initialAttributes = DefaultAttributes.recoverWith
+
+  override def createLogic(attr: Attributes) = new GraphStageLogic(shape) {
+    setHandler(in, new InHandler {
+      override def onPush(): Unit = push(out, grab(in))
+      override def onUpstreamFailure(ex: Throwable) = onFailure(ex)
+    })
+
+    setHandler(out, new OutHandler {
+      override def onPull(): Unit = pull(in)
+    })
+
+    def onFailure(ex: Throwable) = if (pf.isDefinedAt(ex)) switchTo(pf(ex)) else failStage(ex)
+
+    def switchTo(source: Graph[SourceShape[T], M]): Unit = {
+      val sinkIn = new SubSinkInlet[T]("RecoverWithSink")
+      sinkIn.setHandler(new InHandler {
+        override def onPush(): Unit =
+          if (isAvailable(out)) {
+            push(out, sinkIn.grab())
+            sinkIn.pull()
+          }
+        override def onUpstreamFinish(): Unit = if (!sinkIn.isAvailable) completeStage()
```

Comment:
if substream has something than we first emit element downstream and only then complete stage